pkg/eestream/encode.go

```diff
@@ -38,123 +40,309 @@ type ErasureScheme interface {
+			er.cancel()
+		}()
+	}
+	if er.done == er.opt {
+		// optimum threshold reached - cancel the context
+		er.cancel()
 	}
-	// reset piecesRemaining
-	er.piecesRemaining = er.es.TotalCount()
-	return nil
 }
 
 type encodedPiece struct {
-	er *encodedReader
-	i  int
+	ctx    context.Context
+	cancel context.CancelFunc
+	er     *encodedReader
+	ch     chan block
+	outbuf []byte
+	err    error
 }
 
 func (ep *encodedPiece) Read(p []byte) (n int, err error) {
-	// lock! threadsafety matters here
-	ep.er.cv.L.Lock()
-	defer ep.er.cv.L.Unlock()
-
-	outbufs, i := ep.er.outbufs, ep.i
-	if len(outbufs[i]) <= 0 {
-		// if we don't have any buffered result yet, wait until we do
-		err := ep.er.wait()
-		if err != nil {
-			return 0, err
+	if ep.err != nil {
+		return 0, ep.err
+	}
+	if len(ep.outbuf) <= 0 {
+		// take the next block from the channel or block if channel is empty
+		select {
+		case b, ok := <-ep.ch:
+			if !ok {
+				// channel was closed due to slowliness
+				return 0, io.ErrUnexpectedEOF
+			}
+			if b.err != nil {
+				ep.err = b.err
+				if ep.err == io.EOF {
+					ep.er.readerDone()
+				}
+				return 0, ep.err
+			}
+			ep.outbuf = b.data
+		case <-ep.ctx.Done():
+			// context was canceled due to:
+			//  - slowliness
+			//  - optimum threshold reached
+			//  - timeout after reaching minimum threshold expired
+			return 0, io.ErrUnexpectedEOF
 		}
 	}
 
 	// we have some buffer remaining for this piece. write it to the output
-	n = copy(p, outbufs[i])
+	n = copy(p, ep.outbuf)
 	// slide the unused (if any) bytes to the beginning of the buffer
-	copy(outbufs[i], outbufs[i][n:])
+	copy(ep.outbuf, ep.outbuf[n:])
 	// and shrink the buffer
-	outbufs[i] = outbufs[i][:len(outbufs[i])-n]
-	// if there's nothing left, decrement the amount of pieces we have
-	if len(outbufs[i]) <= 0 {
-		ep.er.piecesRemaining--
-	}
+	ep.outbuf = ep.outbuf[:len(ep.outbuf)-n]
 	return n, nil
 }
 
 // EncodedRanger will take an existing Ranger and provide a means to get
 // multiple Ranged sub-Readers. EncodedRanger does not match the normal Ranger
 // interface.
 type EncodedRanger struct {
-	es ErasureScheme
-	rr ranger.Ranger
+	ctx context.Context
+	rr  ranger.Ranger
+	es  ErasureScheme
+	min int // minimum threshold
+	opt int // optimum threshold
+	mbm int // max buffer memory
 }
 
-// NewEncodedRanger creates an EncodedRanger
-func NewEncodedRanger(rr ranger.Ranger, es ErasureScheme) (*EncodedRanger,
-	error) {
+// NewEncodedRanger creates an EncodedRanger. See the comments for EncodeReader
+// about min, opt and mbm.
+func NewEncodedRanger(ctx context.Context, rr ranger.Ranger, es ErasureScheme,
+	min, opt, mbm int) (*EncodedRanger, error) {
 	if rr.Size()%int64(es.DecodedBlockSize()) != 0 {
 		return nil, Error.New("invalid erasure encoder and range reader combo. " +
 			"range reader size must be a multiple of erasure encoder block size")
 	}
+	if min == 0 {
+		min = es.TotalCount()
+	}
+	if opt == 0 {
+		opt = es.TotalCount()
+	}
```

Comment:
no need to check this stuff since EncodeReader does it for you right?

Reply:
It's important to do this before calling `checkParams`. Of course, I can leave it to the EncodeReader, but then any error about invalid input params will be returned by `Range` instead of the constructor.